chainer_chemistry/dataset/parsers/sdf_file_parser.py

```diff
@@ -129,24 +130,12 @@ def parse(self, filepath, retain_smiles=False):
             # Spec not finalized yet for general case
             result = pp.process(filepath)
 
+        smiles_list = numpy.array(smiles_list) if return_smiles else None
         if isinstance(result, tuple):
             if self.postprocess_fn is not None:
                 result = self.postprocess_fn(*result)
-            return NumpyTupleDataset(*result)
+            return NumpyTupleDataset(*result), smiles_list
```

Comment:
It seems weird to me to return `None` as a second element. Apparently it has several drawbacks, for example, because (1) it breaks compatibility and (2) as users know the second element is always `None` when `return_smiles` is `False`, it is meaningless to them. But on the other hand, if we remove this `None`, users need branching in user code depending on the value of `return_smiles`. It seems to me to consider more what is expected specification of parsers.

Reply:
The objective of this PR is changing parsers's behavior. So it can't be avoided breaking compatibility. Your (2) comment is right. Therefore I propose returning dict or namedtuple that contains dataset and smiles from parser.parse. It makes easy to maintain compatibility.

Reply:
OK, understood.